packages/labs/analyzer/src/lib/javascript/modules.ts

```diff
@@ -255,18 +256,29 @@ const getAndValidateModuleFromCache = (
 /**
  * Returns true if all dependencies of the module are still valid.
  */
-const depsAreValid = (module: Module, analyzer: AnalyzerInterface) =>
-  Array.from(module.dependencies).every((path) => depIsValid(path, analyzer));
+const depsAreValid = (
+  module: Module,
+  analyzer: AnalyzerInterface,
+  seen: Set<AbsolutePath>
+): boolean =>
+  Array.from(module.dependencies).every(
+    (path) => seen.has(path) || depIsValid(path, analyzer, seen)
```

Comment:
I'm finding this organization of the logic a little bit hard to reason about, because as used here, `seen.has(path)` implies the dep is valid, but as added in `depIsValid` there is no such assumption. I'm caught wondering if this works correctly, but I don't see it, or if there's a missing test case that shows this doesn't work. Do we need a test with circular imports, but we invalidate one of the modules (but not the first reached in the traversal)?

This may be ok though, but I'd at least like to see a comment about it's safe to short-circuit here, something like "if we've seen this dep and it's not valid we would have already returned false from `depIsValid` for a previous dependent" or similar.

Reply:
How's that?